feat(memberQuiz): 유저-퀴즈 간 기능, dto 수정 및 추가

truuuely opened the pull request:
MemberQuizService에서 redis 사용해 유저가 풀고있는 퀴즈, 푼 갯수 저장

truuuely committed:
refactor(memberQuiz): 디렉토리, 클래스명 수정

truuuely committed:
refactor(quiz): 테이블명, 메서드명 수정

financial_quiz -> quiz

truuuely committed:
feat(memberQuiz): 유저-퀴즈 기능, dto 추가 및 수정

plus100kt commented on src/main/java/com/moayong/api/domain/leaguemember/enums/LeagueMemberErrorCode.java:
```diff
@@ -0,0 +1,14 @@
+@Getter
+@RequiredArgsConstructor
+public enum LeagueMemberErrorCode {
+    LEAGUE_MEMBER_NOT_FOUND(HttpStatus.INTERNAL_SERVER_ERROR, "사용자가 속한 리그를 찾을 수 없습니다");
```
id 로 조회하는데 repository에 없어서 나는 에러면 BAD_REQUEST로 변경

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/config/QuizRedisConfig.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Bean
+    public RedisTemplate<String, Quiz> quizRedisTemplate() {
```
레디스 해쉬로 변경해보자

plus100kt replied:
따로 받는 Repository 가 있어야한다

plus100kt replied:
userId 를 키로 ttl은 내일 오전 9시까지 value는 quiz_id, 필요한 퀴즈 정보, status: "unsolved", "solved"

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/controller/MemberQuizController.java:
```diff
@@ -0,0 +1,59 @@
+    private final MemberQuizService memberQuizService;
+
+    @GetMapping("/users/{id}/quizzes/random")
+    public ApiResponse<DailyQuizResponse> findDailyQUiz(@PathVariable("id") Long userId) {
```
U 대문자 ㅋㅋㅋ

plus100kt replied:
QuizKnowledgeResponse

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+
+        if (quizRedisTemplate.hasKey(dailyQuizKey)) {
+            Quiz cachedQuiz = quizRedisTemplate.opsForValue().get(dailyQuizKey);
+            return new DailyQuizResponse(false, new QuizResponse(cachedQuiz), null);
```
서비스에서 컨트롤러로 이동

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        Long solvedCount = 0L;
+        String solvedCountKey = getSolvedCountKey(userId, criteriaDate);
```
그냥 세는게 좋을듯

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/controller/MemberQuizController.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @GetMapping("/users/{id}/quizzes")
+    public ApiResponse<List<QuizResponse>> findAllSolvedQuizzesByUserId(@PathVariable("id") Long userId) {
```
QuizKnowledgeResponse

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/controller/MemberQuizController.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @GetMapping("/users/{userId}/quizzes/{quizId}/solve")
+    public ApiResponse<QuizViewResponse> showQuiz(@PathVariable("userId") Long userId, @PathVariable("quizId") Long quizId) {
```
GET {quizId}/ploblem
QuizPloblemResponse

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/controller/MemberQuizController.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @PostMapping("/users/{userId}/quizzes/{quizId}/solve")
+    public ApiResponse<QuizSubmissionResponse> submitQuiz(@PathVariable("userId") Long userId, @PathVariable("quizId") Long quizId, @Valid @RequestBody AnswerRequest request) {
```
QuizSubmissionResponse

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/dto/request/AnswerRequest.java:
```diff
@@ -0,0 +1,13 @@
+
+public record AnswerRequest(
+        @Min(1)
+        @Max(3)
```
최소 최대 정해주는애가 있고 없더라도 메시지는 있어야지

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/controller/MemberQuizController.java:
```diff
@@ -0,0 +1,59 @@
+        List<QuizResponse> response = quizzes.stream().map(QuizResponse::new).toList();
+
+        return ApiResponse.success(response, "완료한 퀴즈 전체 조회 성공");
+    }
```
@GetMapping("/users/{id}/quizzes/{id}") 추가

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+
+    public DailyQuizResponse findDailyQuiz(Long userId) {
+        LocalDate criteriaDate = getCriteriaDate();
+        String dailyQuizKey = getDailyQuizKey(userId, criteriaDate);
```
ttl이 있는데 굳이?

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+            solvedCount = (Long) redisTemplate.opsForValue().get(solvedCountKey);
+        }
+
+        if (solvedCount >= 5) {
```
에러

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+
+        Long solvedCount = 0L;
+        String solvedCountKey = getSolvedCountKey(userId, criteriaDate);
+        if (redisTemplate.hasKey(solvedCountKey)) {
```
userid로 매칭된 레코드 개수 가져오기

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+    private Quiz findRandomQuiz(Long userId) {
+        List<Long> solvedQuizIds = memberQuizRepository.findSolvedQuizzesByUserId(userId);
+
+        List<Quiz> quizzes;
```
변수명 풀어야할 퀴즈

plus100kt replied:
get 풀어야할 퀴즈들 함수로 isEmpty까지 처리

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+        LocalDate criteriaDate = getCriteriaDate();
+
+        String dailyQuizKey = getDailyQuizKey(userId, criteriaDate);
+        quizRedisTemplate.delete(dailyQuizKey);
```
ttl 있고 상태 있으니 지우지 말기

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+            redisTemplate.opsForValue().set(solvedCountKey, 0L, getTtlUntilNext9AM());
+        }
+
+        redisTemplate.opsForValue().increment(solvedCountKey, 1L);
```
solvedCountKey 를 가진 함수들 다 지우기

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+
+        // TODO - TotalScore 레디스 처리
+
+        return new QuizSubmissionResponse(userAnswer, status, quiz);
```
이거 컨트롤러단에서 부탁드려요 만약 많은 정보가 필요하면 service - controller dto를 하나 만들기

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    public Quiz findByQuizId(Long quizId) {
+        return quizService.findQuizByIdOptional(quizId)
```
findQuizById를 사용 (에러 처리가 되어있다면)

plus100kt commented on src/main/java/com/moayong/api/domain/quiz/enums/QuizErrorCode.java:
```diff
@@ -8,6 +8,7 @@
 @RequiredArgsConstructor
 public enum QuizErrorCode {
     QUIZ_NOT_FOUND(HttpStatus.INTERNAL_SERVER_ERROR, "퀴즈를 찾을 수 없습니다."),
```
BAD_REQUEST

plus100kt commented on src/main/java/com/moayong/api/domain/quiz/enums/QuizErrorCode.java:
```diff
@@ -8,6 +8,7 @@
 @RequiredArgsConstructor
 public enum QuizErrorCode {
     QUIZ_NOT_FOUND(HttpStatus.INTERNAL_SERVER_ERROR, "퀴즈를 찾을 수 없습니다."),
+    EMPTY_QUIZ_LIST(HttpStatus.INTERNAL_SERVER_ERROR, "퀴즈 목록이 비었습니다."),
```
지우기

plus100kt commented on src/main/java/com/moayong/api/domain/memberQuiz/service/MemberQuizService.java:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        if (quizzes.isEmpty()) {
+            throw new QuizException(QuizErrorCode.EMPTY_QUIZ_LIST);
```
풀어야할 퀴즈가 없어서 에러를 준거니까 MemberQuizException 이 되어야한다

truuuely committed:
feat(memberQuiz): 유저-퀴즈 기능, dto 수정 및 패키지명 수정

- 패키지명 변경: memberQuiz -> memberquiz
- ErrorCode status 수정
- 유저가 당일 푼 퀴즈 Redis 처리 (RedisHash)
- MemberQuiz DTO 수정
